File: src/main/java/waltid/x509/SignCertificateExample.java
package waltid.x509;

import java.util.List;

import id.walt.certificate.x509.JavaX509CertificateUtil;
import id.walt.certificate.x509.X509Certificate;
import id.walt.certificate.x509.extension.AuthorityKeyIdentifierExtension;
import id.walt.certificate.x509.extension.KeyUsageExtension;
import id.walt.certificate.x509.extension.SubjectAlternativeNameExtension;
import id.walt.certificate.x509.extension.SubjectKeyIdentifierExtension;
import id.walt.certificate.x509.truststore.InMemoryTrustStore;
import id.walt.certificate.x509.validation.ValidationResult;
import id.walt.crypto2.CryptoRuntime;
import id.walt.crypto2.algorithms.SignatureAlgorithm;
import id.walt.crypto2.jvm.JavaDigestAlgorithm;
import id.walt.crypto2.jvm.JavaSignatureAlgorithm;
import id.walt.crypto2.jvm.JavaSoftwareKeys;
import id.walt.crypto2.keys.Key;
import kotlin.Unit;

/**
 * Java port of {@code SignCertificateExample.kt}.
 *
 * <p>Creates a self-signed root CA, issues a leaf certificate under it, and validates the chain
 * against the root as trust anchor - showing that neither chain order nor the presence of the
 * root in the chain matters.
 */
public class SignCertificateExample {

    private static final CryptoRuntime cryptoRuntime = JavaSoftwareKeys.defaultRuntime();
    private static final SignatureAlgorithm certSigningAlg =
            JavaSoftwareKeys.toKotlin(JavaSignatureAlgorithm.ecdsa(new JavaDigestAlgorithm("SHA-256"), "DER"));

    public static void main(String[] args) {
        Key caKey = Crypto2Keys.generateEcP256Key(cryptoRuntime, "ca");
        X509Certificate caCert = createSelfSignedRootCa(caKey);
        System.out.println();
        Key leafKey = Crypto2Keys.generateEcP256Key(cryptoRuntime, "ca");
        X509Certificate leafCert = createLeafCertificate(caKey, caCert, leafKey);
        System.out.println();
        validateCertificateChain(List.of(leafCert), caCert);
        System.out.println();
        // including the root in the chain is not necessary but allowed
        validateCertificateChain(List.of(leafCert, caCert), caCert);
        System.out.println();
        // order doesn't matter
        validateCertificateChain(List.of(caCert, leafCert), caCert);
    }

    private static X509Certificate createSelfSignedRootCa(Key key) {
        X509Certificate cert = JavaX509CertificateUtil.getDefault().createSelfSignedCertificate(
                key,
                certSigningAlg,
                builder -> {
                    builder.setSubjectDn("cn=My Root, o=Walt.id, c=AT");
                    // add key usage constraint
                    KeyUsageExtension.Companion.extensionKeyUsage(builder, keyUsage -> {
                        keyUsage.addKeyUsage(
                                KeyUsageExtension.KeyUsage.digitalSignature,
                                KeyUsageExtension.KeyUsage.keyCertSign
                        );
                        return Unit.INSTANCE;
                    });
                    // add subject alternative names
                    SubjectAlternativeNameExtension.Companion.extensionSan(builder, san -> {
                        san.addEmail("example@walt.id");
                        san.addUri("https://walt.id");
                        return Unit.INSTANCE;
                    });
                }
        );
        System.out.println("Created root Ca certificate");
        System.out.println("Issuer DN / Subject DN: " + cert.getData().getIssuerDn());
        SubjectKeyIdentifierExtension ski =
                SubjectKeyIdentifierExtension.Companion.getExtensionSubjectKeyIdentifier(cert.getData());
        System.out.println("Subject Key ID: " + (ski != null ? ski.getKeyIdentifier() : null));
        System.out.println("Fingerprint: " + cert.getFingerprintSha256Hex());
        return cert;
    }

    private static X509Certificate createLeafCertificate(Key issuerKey, X509Certificate issuerCert, Key subjectKey) {
        X509Certificate cert = JavaX509CertificateUtil.getDefault().createCertificate(
                issuerKey,
                issuerCert,
                certSigningAlg,
                builder -> {
                    builder.setSubjectDn("cn=My Leaf Certificate, o=Walt.id, c=AT");
                    builder.subjectPublicKey(subjectKey);
                }
        );
        System.out.println("Created leaf certificate");
        System.out.println("Issuer DN (same as caCert DN): " + cert.getData().getIssuerDn());
        AuthorityKeyIdentifierExtension aki =
                AuthorityKeyIdentifierExtension.Companion.getExtensionAuthorityKeyIdentifier(cert.getData());
        System.out.println("Issuer Key ID (same as subject KeyId of parent Cert): "
                + (aki != null ? aki.getKeyIdentifier() : null));
        System.out.println("Subject DN: " + cert.getData().getSubjectDn());
        System.out.println("Fingerprint: " + cert.getFingerprintSha256Hex());
        // restored subject public key - not used further, mirrors the Kotlin example
        Key publicKey = JavaX509CertificateUtil.getDefault().restoreSubjectPublicKey(cert, cryptoRuntime);
        return cert;
    }

    private static void validateCertificateChain(List<X509Certificate> chain, X509Certificate trustRoot) {
        InMemoryTrustStore trustStore = new InMemoryTrustStore(List.of(trustRoot));
        // Passing trustStore here scopes trust to exactly trustRoot for this call - it replaces
        // X509CertificateUtil.Default's configured trust store rather than merging with it, so this
        // does not also trust the platform's system CA store. See ConfigureTrustStoreExample for
        // more on how trust stores are combined/configured.
        ValidationResult validationResult =
                JavaX509CertificateUtil.getDefault().validateCertificateChain(chain, trustStore);
        System.out.println("Validation result - isValid: " + validationResult.getValid());
        System.out.println("Validation result - log:");
        for (ValidationResult.ValidationLogEntry entry : validationResult.getLog()) {
            System.out.println(entry.getSeverity() + " " + entry.getSubjectDn() + "/"
                    + entry.getValidatorId() + ": '" + entry.getMessage() + "'");
        }
    }
}

File: src/main/kotlin/x509/SignCertificateExample.kt
package x509

import id.walt.certificate.x509.X509Certificate
import id.walt.certificate.x509.X509CertificateUtil
import id.walt.certificate.x509.extension.AuthorityKeyIdentifierExtension.Companion.extensionAuthorityKeyIdentifier
import id.walt.certificate.x509.extension.KeyUsageExtension
import id.walt.certificate.x509.extension.KeyUsageExtension.Companion.extensionKeyUsage
import id.walt.certificate.x509.extension.SubjectAlternativeNameExtension.Companion.extensionSan
import id.walt.certificate.x509.extension.SubjectKeyIdentifierExtension.Companion.extensionSubjectKeyIdentifier
import id.walt.certificate.x509.truststore.InMemoryTrustStore
import id.walt.crypto2.CryptoRuntime
import id.walt.crypto2.algorithms.DigestAlgorithm
import id.walt.crypto2.algorithms.EcdsaSignatureEncoding
import id.walt.crypto2.algorithms.SignatureAlgorithm
import id.walt.crypto2.keys.*
import id.walt.crypto2.providers.GenerateSoftwareKeyRequest
import id.walt.crypto2.providers.cryptography.defaultSoftwareKeyProviders


private val cryptoRuntime = CryptoRuntime(defaultSoftwareKeyProviders())
private val keyGen = GenerateSoftwareKeyRequest(
    id = KeyId("ca"),
    spec = KeySpec.Ec(EcCurve.P256),
    usages = setOf(KeyUsage.SIGN, KeyUsage.VERIFY)
)
private val certSigningAlg = SignatureAlgorithm.Ecdsa(DigestAlgorithm.SHA_256, EcdsaSignatureEncoding.DER)


suspend fun main() {
    val caKey = cryptoRuntime.generateSoftwareKey(keyGen)
    val caCert = createSelfSignedRootCa(caKey)
    println()
    val leafKey = cryptoRuntime.generateSoftwareKey(keyGen)
    val leafCert = createLeafCertificate(caKey, caCert, leafKey)
    println()
    validateCertificateChain(listOf(leafCert), caCert)
    println()
    //including the root in the chain is not necessary but allowed
    validateCertificateChain(listOf(leafCert, caCert), caCert)
    println()
    //order doesn't matter
    validateCertificateChain(listOf(caCert, leafCert), caCert)
}


private suspend fun createSelfSignedRootCa(key: Key): X509Certificate {
    val cert = X509CertificateUtil.createSelfSignedCertificate(key, certSigningAlg) {
        subjectDn = "cn=My Root, o=Walt.id, c=AT"
        //add key usage constraint
        extensionKeyUsage {
            addKeyUsage(KeyUsageExtension.KeyUsage.digitalSignature, KeyUsageExtension.KeyUsage.keyCertSign)
        }
        //add subject alternative names
        extensionSan {
            addEmail("office@walt.id")
            addUri("https://walt.id")
        }
    }
    println("Created root Ca certificate")
    println("Issuer DN / Subject DN: ${cert.data.issuerDn}")
    println("Subject Key ID: ${cert.data.extensionSubjectKeyIdentifier?.keyIdentifier}")
    println("Fingerprint: ${cert.fingerprintSha256Hex}")
    return cert
}

private suspend fun createLeafCertificate(
    issuerKey: Key,
    issuerCert: X509Certificate,
    subjectKey: Key
): X509Certificate {
    val cert = X509CertificateUtil.createCertificate(issuerKey, issuerCert, certSigningAlg) {
        subjectDn = "cn=My Leaf Certificate, o=Walt.id, c=AT"
        subjectPublicKey(subjectKey)
    }
    println("Created leaf certificate")
    println("Issuer DN (same as caCert DN): ${cert.data.issuerDn}")
    println("Issuer Key ID (same as subject KeyId of parent Cert): ${cert.data.extensionAuthorityKeyIdentifier?.keyIdentifier}")
    println("Subject DN: ${cert.data.subjectDn}")
    println("Fingerprint: ${cert.fingerprintSha256Hex}")
    val publicKey = cert.restoreSubjectPublicKey(cryptoRuntime)
    return cert
}

private suspend fun validateCertificateChain(chain: List<X509Certificate>, trustRoot: X509Certificate) {
    val trustStore = InMemoryTrustStore(listOf(trustRoot))
    val validationResult = X509CertificateUtil.validateCertificateChain(chain, trustStore)
    println("Validation result - isValid: ${validationResult.valid}")
    println("Validation result - log:")
    validationResult.log.forEach { println("${it.severity} ${it.subjectDn}/${it.validatorId}: '${it.message}'") }
}